core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -95,6 +95,39 @@ class KafkaController(val config: KafkaConfig,
     new ControllerBrokerRequestBatch(config, controllerChannelManager, eventManager, controllerContext, stateChangeLogger))
   val topicDeletionManager = new TopicDeletionManager(config, controllerContext, replicaStateMachine,
     partitionStateMachine, new ControllerDeletionClient(this, zkClient))
+  val reassignmentListener = new ReassignmentListener {
+
+    // While a reassignment is in progress, deletion is not allowed
+    override def preReassignmentStartOrResume(tp: TopicPartition): Unit =
+      topicDeletionManager.markTopicIneligibleForDeletion(Set(tp.topic), reason = "topic reassignment in progress")
+
+    override def postReassignmentStartedOrResumed(tp: TopicPartition): Unit =
+      zkClient.registerZNodeChangeHandler(new PartitionReassignmentIsrChangeHandler(eventManager, tp))
+
+    override def preReassignmentFinish(tp: TopicPartition, deletedZNode: Boolean): Unit = {
+      val path = TopicPartitionStateZNode.path(tp)
+      zkClient.unregisterZNodeChangeHandler(path)
+      if (deletedZNode) {
+        // Ensure we detect future reassignments
+        eventManager.put(ZkPartitionReassignment)
+      }
+    }
+
+    // signal delete topic thread if reassignment for some partitions belonging to topics being deleted just completed
+    override def postReassignmentFinished(tp: TopicPartition): Unit =
+      topicDeletionManager.resumeDeletionForTopics(Set(tp.topic))
+  }
+  val reassignmentsManager = new ReassignmentManager(controllerContext, zkClient, reassignmentListener,
+    replicaStateMachine, partitionStateMachine, brokerRequestBatch, stateChangeLogger,
+    shouldSkipReassignment = tp => {
```

Comment:
I had one suggestion which you are free to reject. We have this kind of awkward back and forth between the controller and the manager when a reassignment is triggered:

1. Controller detects reassignment
2. Controller delegates trigger to Manager
3. Manager asks Controller if reassignment is allowed
4. Manager executes reassignment

I am wondering if it would be better to leave all of the trigger logic inside the controller and only delegate to the manager after step 3. In other words, perhaps `maybeTriggerPartitionReassignment` can be left inside the controller. What do you think?